In [2]:
# Imports and pip installations (if needed)
import pandas as pd 
import numpy as np 
from scipy import stats
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [3]:
# Load the given datasets
ckdnum = pd.read_csv("/Users/<redacted>/Downloads/chronic_kidney_disease_numerical.csv")
ckdcat = pd.read_csv("/Users/<redacted>/Downloads/chronic_kidney_disease_categorical.csv")

# Print the data
print(ckdnum)
print(ckdcat)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Both datasets are in csv (comma-seperated values) format and contain missing values as indicated by the website where we acquired this data. 
1. chronic_kidney_disease_numerical.csv: The dataset provides the age, blood pressure, blood glucose value and other detailed medical data for patients according to their unique patient ID. This dataset only contains numerical values as indicated by its name and first 5 entries(head).
2. chronic_kidney_disease_categorical.csv: The dataset provides the albumin levels, sugar levels, pus cell clump existence and other detailed medical data for patients according to their unique patient ID. This dataset only contains categorical/non-numerical values as indicated by its name and first 5 entries(head).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [4]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print(ckdnum.duplicated(subset="unique_id").any())
if(ckdnum.duplicated(subset="unique_id").any()):
    print(ckdnum.duplicated(subset="unique_id").sum())

# Drop these duplicate rows
ckdnum_no_dups = ckdnum.drop_duplicates(subset="unique_id")


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print(ckdcat.duplicated(subset="unique_id").any())
if(ckdcat.duplicated(subset="unique_id").any()):
    print(ckdcat.duplicated(subset="unique_id").sum())
print(ckdcat[ckdcat.duplicated(subset="unique_id", keep=False)])
ckdcat_no_dups = ckdcat.drop_duplicates(subset="unique_id")


True
30
True
30
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
10      708255  2.0  0.0    normal    normal  notpresent  notpresent   no   
34      441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
39      240975  0.0  0.0    normal    normal  notpresent  notpresent   no   
51      354223  0.0  0.0    normal    normal  notpresent  notpresent   no   
71      283427  0.0  0.0    normal    normal  notpresent  notpresent   no   
74      302906  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
88      856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
94      501321  1.0  0.0       NaN    normal  notpresent  notpresent   no   
113     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
124     261137  0.0  0.0    normal    normal  notpresent  notpresent   no   
129     643007  NaN  NaN       NaN    normal  notpresent  no

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

All the unique ids are not present in both the datasets. The number of rows in the initial datasets aer oer than that in the combined dataset.

In [5]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
ckd = pd.merge(ckdnum_no_dups, ckdcat_no_dups, on="unique_id")
#Print the combined dataset
print(ckd)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
385     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
386     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
389     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

           pc         pcc          ba  htn   dm cad

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [6]:
# Calculate the percentage of rows that contain atleast one missing value
missing_rows = ckd.isnull().any(axis=1)
num_missing_rows = missing_rows.sum()
percentage_missing = (num_missing_rows / 390) * 100
# Print %
print(percentage_missing)
# Drop these rows from the dataset
ckd_clean =ckd.dropna()
# Print the Dataset
print(ckd_clean)

60.76923076923077
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

           pc         pcc          ba  htn   

## Part 3.4: Sort the dataset according to the Labels

In [7]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_ckd = ckd_clean.sort_values(by='Target')
# Print the dataset
print(sorted_ckd)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
94      137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
106     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
320     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
307     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
131     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
128     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
124     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
173     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
208     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [8]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
ckd_with_dummies = pd.get_dummies(sorted_ckd, columns=categorical_columns)
# Print the dataset
print(ckd_with_dummies)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
94      137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
106     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
320     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
307     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
131     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
128     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
124     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
173     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
208     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [9]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
z_scores = ckd_with_dummies[numerical_columns].apply(stats.zscore)
non_outliers = (abs(z_scores) < 3).all(axis=1)
ckd_standard = ckd_with_dummies[non_outliers]

# Print the dataset
print(ckd_standard)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
94      137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
320     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
307     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
300     546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
131     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
128     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
124     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
173     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
208     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [10]:
# Normalize the all Numerical Attributes in the dataset.
scaler = MinMaxScaler()
ckd_standard[numerical_columns] = scaler.fit_transform(ckd_standard[numerical_columns])
# Print the dataset
print(ckd_standard)

     unique_id       age    bp       bgr        bu        sc       sod  \
388     474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
94      137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
320     343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
307     532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
300     546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
..         ...       ...   ...       ...       ...       ...       ...   
131     647552  0.743243  0.25  0.214592  0.124183  0.037037  0.566667   
128     125587  0.418919  0.00  0.167382  0.241830  0.086420  0.700000   
124     614376  0.689189  0.50  0.210300  0.045752  0.098765  0.500000   
173     853426  0.472973  0.25  0.236052  0.183007  0.024691  0.666667   
208     572140  0.662162  0.50  0.145923  0.117647  0.061728  0.733333   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
388  0.171429  0.221311  0.184211  ...  

/var/folders/9x/hw3zm4hn0db63kdhrpwp6j2h0000gn/T/ipykernel_40158/444404576.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ckd_standard[numerical_columns] = scaler.fit_transform(ckd_standard[numerical_columns])


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: We removed the unique_id column since it is randomly assigned and chronic kideny disease is independant of  patient's unique ID.

In [11]:
#Remove that column
final_ckd = ckd_standard.drop(columns=["unique_id"])
# Print the dataset
print(final_ckd)

          age    bp       bgr        bu        sc       sod       pot  \
388  0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.171429   
94   0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
320  0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.657143   
307  0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.657143   
300  0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.314286   
..        ...   ...       ...       ...       ...       ...       ...   
131  0.743243  0.25  0.214592  0.124183  0.037037  0.566667  0.171429   
128  0.418919  0.00  0.167382  0.241830  0.086420  0.700000  0.571429   
124  0.689189  0.50  0.210300  0.045752  0.098765  0.500000  0.514286   
173  0.472973  0.25  0.236052  0.183007  0.024691  0.666667  0.600000   
208  0.662162  0.50  0.145923  0.117647  0.061728  0.733333  0.542857   

         hemo       pcv      wbcc  ...  cad_no  cad_yes  appet_good  \
388  0.221311  0.184211  0.653226  ...    True    Fa

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [12]:
# Export the dataframe to a new csv file
final_ckd.to_csv('chronic_kidney_disease_data.csv', index = False)
# Export the dataframe to a new json file
final_ckd.to_json('chronic_kidney_disease_data.json', index = False)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

| age       | bp   | bgr                  | bu                  | sc                 | sod                 | pot                 | hemo              | pcv               | wbcc              | rbcc              | al   | su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----------|------|----------------------|---------------------|--------------------|---------------------|---------------------|--------------------|-------------------|-------------------|-------------------|------|----|--------------|------------|-------------|-----------|----------------|-------------|----------------|------------|--------|---------|-------|--------|--------|---------|------------|------------|-------|--------|--------|---------|-------------|----------------|
| 0.202703  | 0.75 | 0.15879828326180256 | 0.19607843137254904 | 0.16049382716049385 | 0.16666666666666696 | 0.17142857142857149 | 0.2213114754098361 | 0.1842105263157895 | 0.6532258064516129 | 0.33333333333333326 | 4.0  | 0.0| False        | True       | True        | False     | False          | True        | False          | True       | True   | False   | True  | False  | True   | False   | True        | False      | True   | False  | True   | False   | True        | False          |
| 0.905405  | 1.0  | 0.9656652360515021  | 0.5228758169934641  | 0.6419753086419753  | 0.666666666666667   | 0.0                | 0.29508196721311464| 0.368421052631579 | 0.217741935483871  | 0.15384615384615397 | 3.0  | 2.0| True         | False      | True        | False     | False          | True        | True           | False      | False  | True    | False | True   | False  | True    | False       | True       | False  | True   | False  | True    | True        | False          |
| 0.743243  | 0.5  | 0.44206008583690987 | 0.9019607843137255  | 0.4320987654320988  | 0.5                 | 0.6571428571428573  | 0.17213114754098358| 0.21052631578947367| 0.3951612903225806 | 0.15384615384615397 | 2.0  | 0.0| True         | False      | True        | False     | True           | False       | True           | False      | False  | True    | False | True   | False  | True    | False       | True       | True   | False  | True   | False   | True        | False          |
| 0.729730  | 0.75 | 0.1502145922746781  | 0.28104575163398693 | 0.2345679012345679  | 0.5333333333333332  | 0.6571428571428573  | 0.4508196721311475 | 0.4473684210526315 | 0.5               | 0.3846153846153846 | 2.0  | 0.0| True         | False      | False       | True      | True           | False       | True           | False      | True   | False   | True  | False  | True   | False   | True        | False      | True   | False  | True   | False   | True        | False          |
| 0.540541  | 0.0  | 0.39914163090128757 | 0.5359477124183006  | 0.35802469135802467 | 0.7000000000000002  | 0.3142857142857143  | 0.34426229508196726| 0.3157894736842105 | 0.8306451612903226 | 0.15384615384615397 | 1.0  | 0.0| False        | True       | False       | True      | True           | False       | True           | False      | False  | True    | False | True   | False  | True    | False       | True       | True   | False  | True   | False   | True        | False          |
| 0.675676  | 0.75 | 0.25321888412017163 | 0.6339869281045752  | 0.7777777777777779  | 0.36666666666666625 | 0.5428571428571427  | 0.28688524590163933| 0.3421052631578947 | 0.16935483870967738 | 0.20512820512820518 | 2.0  | 0.0| True         | False      | True        | False     | True           | False       | True           | False      | False  | True    | False | True   | False  | True    | False       | True       | True   | False  | True   | False   | True        | False          |
| 0.716216  | 0.5  | 1.0                 | 0.16339869281045755 | 0.11111111111111113 | 0.06666666666666643 | 0.17142857142857149 | 0.39344262295081966| 0.5               | 0.532258064516129  | 0.4358974358974359 | 1.0  | 0.0| True         | False      | False       | True      | True           | False       | True           | False      | True   | False   | True  | False  | True   | False   | True        | False      | True   | False  | True   | False   | True        | False          |
| 0.729730  | 0.0  | 0.9356223175965666  | 0.16993464052287582 | 0.16049382716049385 | 0.33333333333333304 | 0.02857142857142858 | 0.1885245901639344 | 0.23684210526315785| 0.8790322580645161 | 0.10256410256410264 | 3.0  | 1.0| False        | True       | True        | False     | False          | True        | True           | False      | False  | True    | True  | False  | True   | False   | True        | False      | True   | False  | True   | False   | True        | False          |
| 0.837838  | 0.5  | 0.3733905579399141  | 0.5228758169934641  | 0.45679012345679015 | 0.33333333333333304 | 1.0                | 0.0               | 0.0               | 0.5403225806451613 | 0.0               | 3.0  | 2.0| False        | True       | True        | False     | False          | True        | False          | True       | False  | True    | False | True   | False  | True    | False       | True       | False  | True   | False  | True    | True        | False          |
| 0.0      | 0.0  | 0.10300429184549353 | 0.37254901960784315 | 0.07407407407407407 | 0.5                 | 0.5714285714285716  | 0.3524590163934427 | 0.368421052631579 | 0.9999999999999999 | 0.5641025641025641 | 4.0  | 0.0| True         | False      | True        | False     | True           | False       | False          | True       | True   | False   | True  | False  | True   | False   | True        | False      | True   | False  | True   | False   | True        | False          |
| 0.878378  | 0.0  | 0.20600858369098712 | 0.7516339869281046  | 0.6049382716049383  | 0.5333333333333332  | 0.5714285714285716  | 0.47540983606557374| 0.5               | 0.8790322580645161 | 0.4358974358974359 | 4.0  | 0.0| False        | True       | False       | True      | True           | False       | True           | False      | True   | False   | True  | False  | True   | False   | True        | False      | True   | False  | True   | False   | True        | False          |
| 0.851351  | 0.25 | 0.6180257510729614  | 0.5620915032679739  | 0.7283950617283951  | 0.0                 | 0.2857142857142856  | 0.3114754098360656 | 0.3157894736842105 | 0.5806451612903225 | 0.17948717948717952 | 4.0  | 3.0| False        | True       | True        | False     | False          | True        | False          | True       | False  | True    | False | True   | False  | True    | False       | True       | False  | True   | False  | True    | True        | False          |
| 0.878378  | 0.25 | 0.6394849785407726  | 0.47058823529411764 | 0.39506172839506176 | 0.43333333333333357 | 0.4285714285714286  | 0.39344262295081966| 0.4473684210526315 | 0.10483870967741932 | 0.2564102564102565 | 3.0  | 0.0| False        | True       | True        | False     | False          | True        | False          | True       | False  | True    | False | True   | False  | True    | False       | True       | False  | True   | False  | True    | True        | False          |
| 0.783784  | 0.0  | 0.7253218884120172  | 0.3137254901960784  | 0.4814814814814815  | 0.5666666666666664  | 0.7142857142857144  | 0.319672131147541  | 0.3421052631578947 | 0.25806451612903225 | 0.20512820512820518 | 4.0  | 1.0| True         | False      | True        | False     | True           | False       | False          | True       | False  | True    | False | True   | False  | True    | False       | True       | True   | False  | True   | False   | True        | False          |
| 0.662162  | 0.5  | 0.6180257510729614  | 0.411764705882353   | 0.4320987654320988  | 0.5666666666666664  | 0.5714285714285716  | 0.43442622950819676| 0.4736842105263157 | 0.25              | 0.28205128205128216 | 3.0  | 1.0| False        | True       | True        | False     | False          | True        | False          | True       | False  | True    | False | True   | False  | True    | False       | True       | False  | True   | False  | True    | True        | False          |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

1. Cross-Check with Source Data: After the conversion, always compare the output with the source data. Implement consistency checks to ensure that the converted data retains the same values and structure as the original. 
2. Automated Testing: Create validation scripts that can automatically check key statistics (e.g., counts, means, medians) before and after conversion to ensure that the essential characteristics of the dataset have not changed. 
3. Small Batches: Instead of converting large datasets all at once, break the data into smaller batches. Convert and validate each batch before proceeding to the next. This makes it easier to identify specific areas where hallucinations occur.


## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [13]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rbcc': 'red blood cell count',
    'al': 'albumin',
    'su': 'sugar',
    'rbc_abnormal': 'red blood cells abnormal',
    'rbc_normal': 'red blood cells normal',
    'pc_abnormal': 'pus cell abnormal',
    'pc_normal': 'pus cell normal',
    'pcc_notpresent': 'pus cell clumps not present',
    'pcc_present': 'pus cell clumps present',
    'ba_notpresent': 'bacteria not present',
    'ba_present': 'bacteria present',
    'htn_no': 'hypertension no',
    'htn_yes': 'hypertension yes',
    'dm_no': 'diabetes mellitus no',
    'dm_yes': 'diabetes mellitus yes',
    'cad_no': 'coronary artery disease no',
    'cad_yes': 'coronary artery disease yes',
    'appet_good': 'appetite good',
    'appet_poor': 'appetite poor',
    'pe_no': 'pedal edema no',
    'pe_yes': 'pedal edema yes',
    'ane_no': 'anemia no',
    'ane_yes': 'anemia yes',
    'Target_ckd': 'ckd',
    'Target_notckd': 'not ckd',
}

final_ckd.rename(columns=rename_mapping, inplace=True)


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [14]:
# Converted SQL to Pandas code
result = final_ckd.groupby('ckd').size().reset_index(name='count')
print(result)

     ckd  count
0  False    111
1   True     23


*Describe what the above code does here*
1. Groups the DataFrame by the "Target_ckd" column.
2. Computes the count of rows for each group.
3. Resets the index of the resulting Series to convert it back to a DataFrame, naming the count column as "count".
4. Resulting Output: The result DataFrame will contain two columns: "Target_ckd" (unique values from the original "Target" column) and "count" (number of occurrences of each unique value).